File: cpp/src/Ice/OpaqueEndpointI.cpp
// **********************************************************************
//
// Copyright (c) 2003-2013 ZeroC, Inc. All rights reserved.
//
// This copy of Ice is licensed to you under the terms described in the
// ICE_LICENSE file included in this distribution.
//
// **********************************************************************

#include <Ice/OpaqueEndpointI.h>
#include <Ice/BasicStream.h>
#include <Ice/Exception.h>
#include <Ice/DefaultsAndOverrides.h>
#include <Ice/Base64.h>
#include <Ice/HashUtil.h>
#include <Ice/LocalException.h>

using namespace std;
using namespace Ice;
using namespace IceInternal;

IceInternal::OpaqueEndpointI::OpaqueEndpointI(vector<string>& args) :
    EndpointI(""), _rawEncoding(Encoding_1_0)
{
    vector<string> unknown;

    ostringstream ostr;
    for(vector<string>::iterator p = args.begin(); p != args.end(); ++p)
    {
        if(p->find_first_of(" \t\n\r") != string::npos)
        {
            ostr << " \"" << *p << "\"";
        }
        else
        {
            ostr << " " << *p;
        }
    }
    const string str = ostr.str();

    int topt = 0;
    int vopt = 0;

    for(vector<string>::size_type n = 0; n < args.size(); ++n)
    {
        string option = args[n];
        if(option.length() < 2 || option[0] != '-')
        {
            unknown.push_back(option);
            continue;
        }

        string argument;
        if(n + 1 < args.size() && args[n + 1][0] != '-')
        {
            argument = args[++n];
        }

        switch(option[1])
        {
            case 't':
            {
                if(argument.empty())
                {
                    EndpointParseException ex(__FILE__, __LINE__);
                    ex.str = "no argument provided for -t option in endpoint `opaque" + str + "'";
                    throw ex;
                }
                istringstream p(argument);
                Ice::Int t;
                if(!(p >> t) || !p.eof())
                {
                    EndpointParseException ex(__FILE__, __LINE__);
                    ex.str = "invalid type value `" + argument + "' in endpoint `opaque" + str + "'";
                    throw ex;
                }
                else if(t < 0 || t > 65535)
                {
                    EndpointParseException ex(__FILE__, __LINE__);
                    ex.str = "type value `" + argument + "' out of range in endpoint `opaque" + str + "'";
                    throw ex;
                }
                _type = static_cast<Ice::Short>(t);
                ++topt;
                if(topt > 1)
                {
                    EndpointParseException ex(__FILE__, __LINE__);
                    ex.str = "multiple -t options in endpoint `opaque" + str + "'";
                    throw ex;
                }
                break;
            }

            case 'v':
            {
                if(argument.empty())
                {
                    EndpointParseException ex(__FILE__, __LINE__);
                    ex.str = "no argument provided for -v option in endpoint `opaque" + str + "'";
                    throw ex;
                }
                for(string::size_type i = 0; i < argument.size(); ++i)
                {
                    if(!Base64::isBase64(argument[i]))
                    {
                        EndpointParseException ex(__FILE__, __LINE__);
                        ostringstream ostr;
                        ostr << "invalid base64 character `" << argument[i] << "' (ordinal " << (int)argument[i]
                             << ") in endpoint `opaque" << str << "'";
                        ex.str = ostr.str();
                        throw ex;
                    }
                }
                const_cast<vector<Byte>&>(_rawBytes) = Base64::decode(argument);
                ++vopt;
                if(vopt > 1)
                {
                    EndpointParseException ex(__FILE__, __LINE__);
                    ex.str = "multiple -v options in endpoint `opaque" + str + "'";
                    throw ex;
                }
                break;
            }

            case 'e':
            {
                if(argument.empty())
                {
                    Ice::EndpointParseException ex(__FILE__, __LINE__);
                    ex.str = "no argument provided for -e option in endpoint `opaque" + str + "'";
                    throw ex;
                }

                try
                {
                    _rawEncoding = Ice::stringToEncodingVersion(argument);
                }
                catch(const Ice::VersionParseException& e)
                {
                    Ice::EndpointParseException ex(__FILE__, __LINE__);
                    ex.str = "invalid encoding version `" + argument + "' in endpoint `opaque" + str + "':\n" + e.str;
                    throw ex;
                }
                break;
            }

            default:
            {
                unknown.push_back(option);
                break;
            }
        }
    }

    if(topt != 1)
    {
        EndpointParseException ex(__FILE__, __LINE__);
        ex.str = "no -t option in endpoint `opaque" + str + "'";
        throw ex;
    }
    if(vopt != 1)
    {
        EndpointParseException ex(__FILE__, __LINE__);
        ex.str = "no -v option in endpoint `opaque" + str + "'";
        throw ex;
    }

    //
    // Replace argument vector with only those we didn't recognize.
    //
    args = unknown;
}

IceInternal::OpaqueEndpointI::OpaqueEndpointI(Short type, BasicStream* s) :
    EndpointI(""),
    _type(type)
{
    _rawEncoding = s->getReadEncoding();
    Int sz = s->getReadEncapsSize();
    s->readBlob(const_cast<vector<Byte>&>(_rawBytes), sz);
}

void
IceInternal::OpaqueEndpointI::startStreamWrite(BasicStream* s) const
{
    s->startWriteEncaps(_rawEncoding, DefaultFormat);
}

void
IceInternal::OpaqueEndpointI::streamWrite(BasicStream* s) const
{
    s->writeBlob(_rawBytes);
}

namespace
{

class InfoI : public Ice::OpaqueEndpointInfo
{
public:

    InfoI(Ice::Short type, const Ice::EncodingVersion& rawEncoding, const Ice::ByteSeq& rawByes);

    virtual Ice::Short
    type() const
    {
        return _type;
    }

    virtual bool
    datagram() const
    {
        return false;
    }

    virtual bool
    secure() const
    {
        return false;
    }

private:

    Ice::Short _type;
};


//
// COMPILERFIX: inlining this constructor causes crashes with gcc 4.0.1.
//
InfoI::InfoI(Ice::Short type, const Ice::EncodingVersion& rawEncoding, const Ice::ByteSeq& rawBytes) :
    Ice::OpaqueEndpointInfo(-1, false, rawEncoding, rawBytes),
    _type(type)
{
}

}

Ice::EndpointInfoPtr
IceInternal::OpaqueEndpointI::getInfo() const
{
    return new InfoI(_type, _rawEncoding, _rawBytes);
}

Short
IceInternal::OpaqueEndpointI::type() const
{
    return _type;
}

std::string
IceInternal::OpaqueEndpointI::protocol() const
{
    return "opaque";
}

Int
IceInternal::OpaqueEndpointI::timeout() const
{
    return -1;
}

EndpointIPtr
IceInternal::OpaqueEndpointI::timeout(Int) const
{
    return const_cast<OpaqueEndpointI*>(this);
}

EndpointIPtr
IceInternal::OpaqueEndpointI::connectionId(const string&) const
{
    return const_cast<OpaqueEndpointI*>(this);
}

bool
IceInternal::OpaqueEndpointI::compress() const
{
    return false;
}

EndpointIPtr
IceInternal::OpaqueEndpointI::compress(bool) const
{
    return const_cast<OpaqueEndpointI*>(this);
}

bool
IceInternal::OpaqueEndpointI::datagram() const
{
    return false;
}

bool
IceInternal::OpaqueEndpointI::secure() const
{
    return false;
}

TransceiverPtr
IceInternal::OpaqueEndpointI::transceiver(EndpointIPtr& endp) const
{
    endp = const_cast<OpaqueEndpointI*>(this);
    return 0;
}

vector<ConnectorPtr>
IceInternal::OpaqueEndpointI::connectors(Ice::EndpointSelectionType) const
{
    vector<ConnectorPtr> ret;
    return ret;
}

void
IceInternal::OpaqueEndpointI::connectors_async(Ice::EndpointSelectionType, const EndpointI_connectorsPtr& cb) const
{
    cb->connectors(vector<ConnectorPtr>());
}

AcceptorPtr
IceInternal::OpaqueEndpointI::acceptor(EndpointIPtr& endp, const string&) const
{
    endp = const_cast<OpaqueEndpointI*>(this);
    return 0;
}

vector<EndpointIPtr>
IceInternal::OpaqueEndpointI::expand() const
{
    vector<EndpointIPtr> endps;
    endps.push_back(const_cast<OpaqueEndpointI*>(this));
    return endps;
}

bool
IceInternal::OpaqueEndpointI::equivalent(const EndpointIPtr&) const
{
    return false;
}

bool
IceInternal::OpaqueEndpointI::operator==(const LocalObject& r) const
{
    const OpaqueEndpointI* p = dynamic_cast<const OpaqueEndpointI*>(&r);
    if(!p)
    {
        return false;
    }

    if(this == p)
    {
        return true;
    }

    if(_type != p->_type)
    {
        return false;
    }

    if(_rawEncoding != p->_rawEncoding)
    {
        return false;
    }

    if(_rawBytes != p->_rawBytes)
    {
        return false;
    }

    return true;
}

bool
IceInternal::OpaqueEndpointI::operator<(const LocalObject& r) const
{
    const OpaqueEndpointI* p = dynamic_cast<const OpaqueEndpointI*>(&r);
    if(!p)
    {
        const EndpointI* e = dynamic_cast<const EndpointI*>(&r);
        if(!e)
        {
            return false;
        }
        return type() < e->type();
    }

    if(this == p)
    {
        return false;
    }

    if(_type < p->_type)
    {
        return true;
    }
    else if(p->_type < _type)
    {
        return false;
    }

    if(_rawEncoding < p->_rawEncoding)
    {
        return true;
    }
    else if(p->_rawEncoding < _rawEncoding)
    {
        return false;
    }

    if(_rawBytes < p->_rawBytes)
    {
        return true;
    }
    else if(p->_rawBytes < _rawBytes)
    {
        return false;
    }

    return false;
}

void
IceInternal::OpaqueEndpointI::hashInit(Ice::Int& h) const
{
    hashAdd(h, _rawEncoding.major);
    hashAdd(h, _rawEncoding.minor);
    hashAdd(h, _rawBytes);
}

string
IceInternal::OpaqueEndpointI::options() const
{
    ostringstream s;
    string val = Base64::encode(_rawBytes);
    s << " -t " << _type << " -e " << _rawEncoding << " -v " << val;
    return s.str();
}

File: js/src/Ice/OpaqueEndpointI.js
// **********************************************************************
//
// Copyright (c) 2003-2013 ZeroC, Inc. All rights reserved.
//
// This copy of Ice is licensed to you under the terms described in the
// ICE_LICENSE file included in this distribution.
//
// **********************************************************************

var Base64 = require("./Base64").Ice.Base64;
var Debug = require("./Debug").Ice.Debug;
var HashUtil = require("./HashUtil").Ice.HashUtil;
var Protocol = require("./Protocol").Ice.Protocol;
var StringUtil = require("./StringUtil").Ice.StringUtil;

var _merge = require("Ice/Util").merge;

var Ice = {};
_merge(Ice, require("./Endpoint").Ice);
_merge(Ice, require("./LocalException").Ice);

var OpaqueEndpointI = function()
{
}

OpaqueEndpointI.prototype = new Ice.Endpoint();
OpaqueEndpointI.prototype.constructor = OpaqueEndpointI;

OpaqueEndpointI.fromString = function(str)
{
    var result = new OpaqueEndpointI();

    result._rawEncoding = Protocol.Encoding_1_0;

    var topt = 0;
    var vopt = 0;

    var arr = str.split(/[ \t\n\r]+/);
    var i = 0;
    while(i < arr.length)
    {
        if(arr[i].length == 0)
        {
            i++;
            continue;
        }

        var option = arr[i++];
        if(option.length != 2 && option.charAt(0) != '-')
        {
            throw new Ice.EndpointParseException("expected an endpoint option but found `" + option +
                                                     "' in endpoint `opaque " + str + "'");
        }

        var argument = null;
        if(i < arr.length && arr[i].charAt(0) != '-')
        {
            argument = arr[i++];
        }

        switch(option.charAt(1))
        {
            case 't':
            {
                if(argument === null)
                {
                    throw new Ice.EndpointParseException("no argument provided for -t option in endpoint `opaque "
                                                             + str + "'");
                }

                var type;

                try
                {
                    type = StringUtil.toInt(argument);
                }
                catch(ex)
                {
                    throw new Ice.EndpointParseException("invalid type value `" + argument +
                                                             "' in endpoint `opaque " + str + "'");
                }

                if(type < 0 || type > 65535)
                {
                    throw new Ice.EndpointParseException("type value `" + argument +
                                                             "' out of range in endpoint `opaque " + str + "'");
                }

                result._type = type;
                ++topt;
                if(topt > 1)
                {
                    throw new Ice.EndpointParseException("multiple -t options in endpoint `opaque " + str + "'");
                }
                break;
            }

            case 'v':
            {
                if(argument === null || argument.length === 0)
                {
                    throw new Ice.EndpointParseException("no argument provided for -v option in endpoint `opaque "
                                                             + str + "'");
                }
                for(var j = 0; j < argument.length; ++j)
                {
                    if(!Base64.isBase64(argument.charAt(j)))
                    {
                        throw new Ice.EndpointParseException("invalid base64 character `" + argument.charAt(j) +
                                                                 "' (ordinal " + argument.charCodeAt(j) +
                                                                 ") in endpoint `opaque " + str + "'");
                    }
                }
                result._rawBytes = Base64.decode(argument);
                ++vopt;
                if(vopt > 1)
                {
                    throw new Ice.EndpointParseException("multiple -v options in endpoint `opaque " + str + "'");
                }
                break;
            }

            case 'e':
            {
                if(argument === null)
                {
                    throw new Ice.EndpointParseException("no argument provided for -e option in endpoint `opaque "
                                                             + str + "'");
                }
                try
                {
                    result._rawEncoding = Protocol.stringToEncodingVersion(argument);
                }
                catch(e)
                {
                    throw new Ice.EndpointParseException("invalid encoding version `" + argument +
                                                             "' in endpoint `opaque " + str + "':\n" + e.str);
                }
                break;
            }

            default:
            {
                throw new Ice.EndpointParseException("invalid option `" + option + "' in endpoint `opaque " +
                                                         str + "'");
            }
        }
    }

    if(topt != 1)
    {
        throw new Ice.EndpointParseException("no -t option in endpoint `opaque " + str + "'");
    }
    if(vopt != 1)
    {
        throw new Ice.EndpointParseException("no -v option in endpoint `opaque " + str + "'");
    }
    result.calcHashValue();
    return result;
}

OpaqueEndpointI.fromStream = function(type, s)
{
    var result = new OpaqueEndpointI();
    result._type = type;
    result._rawEncoding = s.startReadEncaps();
    var sz = s.getReadEncapsSize();
    result._rawBytes = s.readBlob(sz);
    s.endReadEncaps();
    result.calcHashValue();
    return result;
}

//
// Marshal the endpoint
//
OpaqueEndpointI.prototype.streamWrite = function(s)
{
    s.writeShort(this._type);
    s.startWriteEncaps(this._rawEncoding, DefaultFormat); // TODO
    s.writeBlob(this._rawBytes);
    s.endWriteEncaps();
}

//
// Convert the endpoint to its string form
//
OpaqueEndpointI.prototype.toString = function()
{
    var val = Base64.encode(this._rawBytes);
    return "opaque -t " + this._type + " - e " + Protocol.encodingVersionToString(this._rawEncoding) + " -v " + val;
}

//
// Return the endpoint information.
//
OpaqueEndpointI.prototype.getInfo = function()
{
    return new OpaqueEndpointInfoI(-1, false, this._rawEncoding, this._rawBytes, this._type);
}

//
// Return the endpoint type
//
OpaqueEndpointI.prototype.type = function()
{
    return this._type;
}

OpaqueEndpointI.prototype.protocol = function()
{
    return "opaque";
}

//
// Return the timeout for the endpoint in milliseconds. 0 means
// non-blocking, -1 means no timeout.
//
OpaqueEndpointI.prototype.timeout = function()
{
    return -1;
}

//
// Return a new endpoint with a different timeout value, provided
// that timeouts are supported by the endpoint. Otherwise the same
// endpoint is returned.
//
OpaqueEndpointI.prototype.changeTimeout = function(t)
{
    return this;
}

//
// Return a new endpoint with a different connection id.
//
OpaqueEndpointI.prototype.changeConnectionId = function(connectionId)
{
    return this;
}

//
// Return true if the endpoints support bzip2 compress, or false
// otherwise.
//
OpaqueEndpointI.prototype.compress = function()
{
    return false;
}

//
// Return a new endpoint with a different compression value,
// provided that compression is supported by the
// endpoint. Otherwise the same endpoint is returned.
//
OpaqueEndpointI.prototype.changeCompress = function(compress)
{
    return this;
}

//
// Return true if the endpoint is datagram-based.
//
OpaqueEndpointI.prototype.datagram = function()
{
    return false;
}

//
// Return true if the endpoint is secure.
//
OpaqueEndpointI.prototype.secure = function()
{
    return false;
}

//
// Get the encoded endpoint.
//
OpaqueEndpointI.prototype.rawBytes = function()
{
    return this._rawBytes; // Returns a Uint8Array
}

//
// Return a server side transceiver for this endpoint, or null if a
// transceiver can only be created by an acceptor. In case a
// transceiver is created, this operation also returns a new
// "effective" endpoint, which might differ from this endpoint,
// for example, if a dynamic port number is assigned.
//
OpaqueEndpointI.prototype.transceiver = function(endpoint)
{
    endpoint.value = null;
    return null;
}

//
// Return an acceptor for this endpoint, or null if no acceptors
// is available. In case an acceptor is created, this operation
// also returns a new "effective" endpoint, which might differ
// from this endpoint, for example, if a dynamic port number is
// assigned.
//
OpaqueEndpointI.prototype.acceptor = function(endpoint, adapterName)
{
    endpoint.value = this;
    return null;
}

OpaqueEndpointI.prototype.connect = function()
{
    return null;
}

//
// Check whether the endpoint is equivalent to another one.
//
OpaqueEndpointI.prototype.equivalent = function(endpoint)
{
    return false;
}

OpaqueEndpointI.prototype.hashCode = function()
{
    return this._hashCode;
}

//
// Compare endpoints for sorting purposes
//
OpaqueEndpointI.prototype.equals = function(p)
{
    if(!(p instanceof OpaqueEndpointI))
    {
        return false;
    }

    if(this === p)
    {
        return true;
    }

    if(this._type !== p._type)
    {
        return false;
    }

    if(!this._rawEncoding.equals(p._rawEncoding))
    {
        return false;
    }

    if(this._rawBytes.length !== p._rawBytes.length)
    {
        return false;
    }
    for(var i = 0; i < this._rawBytes.length; i++)
    {
        if(this._rawBytes[i] !== p._rawBytes[i])
        {
            return false;
        }
    }

    return true;
}

OpaqueEndpointI.prototype.compareTo = function(p)
{
    if(this === p)
    {
        return 0;
    }

    if(p === null)
    {
        return 1;
    }

    if(!(p instanceof OpaqueEndpointI))
    {
        return this.type() < p.type() ? -1 : 1;
    }

    if(this._type < p._type)
    {
        return -1;
    }
    else if(p._type < this._type)
    {
        return 1;
    }

    if(this._rawEncoding.major < p._rawEncoding.major)
    {
        return -1;
    }
    else if(p._rawEncoding.major < this._rawEncoding.major)
    {
        return 1;
    }

    if(this._rawEncoding.minor < p._rawEncoding.minor)
    {
        return -1;
    }
    else if(p._rawEncoding.minor < this._rawEncoding.minor)
    {
        return 1;
    }

    if(this._rawBytes.length < p._rawBytes.length)
    {
        return -1;
    }
    else if(p._rawBytes.length < this._rawBytes.length)
    {
        return 1;
    }
    for(var i = 0; i < this._rawBytes.length; i++)
    {
        if(this._rawBytes[i] < p._rawBytes[i])
        {
            return -1;
        }
        else if(p._rawBytes[i] < this._rawBytes[i])
        {
            return 1;
        }
    }

    return 0;
}

OpaqueEndpointI.prototype.calcHashValue = function()
{
    var h = 5381;
    h = HashUtil.addNumber(h, this._type);
    h = HashUtil.addHashable(h, this._rawEncoding);
    h = HashUtil.addNumberArray(h, this._rawBytes);
    this._hashCode = h;
}

module.exports.Ice = {};
module.exports.Ice.OpaqueEndpointI = OpaqueEndpointI;

var OpaqueEndpointInfoI = function(timeout, compress, rawEncoding, rawBytes, type)
{
    Ice.OpaqueEndpointInfo.call(this, -1, false, rawEncoding, rawBytes);
    this._type = type;
}

OpaqueEndpointInfoI.prototype = new Ice.OpaqueEndpointInfo();
OpaqueEndpointInfoI.prototype.constructor = OpaqueEndpointInfoI();

OpaqueEndpointInfoI.prototype.type = function()
{
    return this._type;
}

OpaqueEndpointInfoI.prototype.datagram = function()
{
    return false;
}

OpaqueEndpointInfoI.prototype.secure = function()
{
    return false;
}
